File: python/integrals.py
import numpy as np
import math

# Code up the integral equations
def compute_S_primitive(alpha, beta, R_A, R_B):
    """Compute the overlap integral between two primitive Gaussians"""
    #Normalisation of s-type Gaussians
    N_alpha = (2.0 * alpha / np.pi)**(0.75)
    N_beta  = (2.0 * beta  / np.pi)**(0.75)

    S_unnormalised =  (np.pi / (alpha + beta))**(3/2) * np.exp(-alpha * beta / (alpha + beta) * np.sum((R_A - R_B)**2))

    return N_alpha * N_beta * S_unnormalised

def compute_T_primitive(alpha, beta, R_A, R_B, S_prim):
    """Compute the kinetic energy integral between two primitive Gaussians"""
    reduced_exp = alpha * beta / (alpha + beta)
    return reduced_exp * (3 - 2 * reduced_exp * np.sum((R_A - R_B)**2)) * S_prim

def compute_V_nuc_primitive(alpha, beta, R_A, R_B, R_nuc):
    """
    Compute nuclear attraction integral for a single nucleus
    <chi_a | -1/|r-R_nuc| | chi_b>
    """
    # Normalization
    N_alpha = (2.0 * alpha / np.pi)**(0.75)
    N_beta = (2.0 * beta / np.pi)**(0.75)
    
    zeta = alpha + beta
    P = (alpha * R_A + beta * R_B) / zeta
    
    # Gaussian product overlap factor
    AB_sq = np.sum((R_A - R_B)**2)
    K_AB = np.exp(-alpha * beta / zeta * AB_sq)
    
    # Distance from product center to nucleus
    PC_sq = np.sum((P - R_nuc)**2)
    
    if PC_sq < 1e-10:
        V_unnormalised = 2.0 * np.pi / zeta  
    else:
        T = zeta * PC_sq
        F0 = 0.5 * np.sqrt(np.pi / T) * math.erf(np.sqrt(T))
        V_unnormalised = 2.0 * np.pi / zeta * F0 
    
    V_prim = N_alpha * N_beta * K_AB * V_unnormalised
    
    return V_prim

def build_S_and_T_matrices(basis_functions):
    """Build the overlap matrix"""
    n_basis = len(basis_functions)
    S = np.zeros((n_basis, n_basis))
    T = np.zeros((n_basis, n_basis))

    for i, basis_i in enumerate(basis_functions):
        for j, basis_j in enumerate(basis_functions):
            R_A = basis_i['center']
            R_B = basis_j['center']

            # Loop over every primitve for the contracted Gaussians
            for alpha, coeff_A in zip(basis_i['exponents'], basis_i['coefficients']):
                for beta, coeff_B in zip(basis_j['exponents'], basis_j['coefficients']):

                    S_prim = compute_S_primitive(alpha, beta, R_A, R_B)
                    T_prim = compute_T_primitive(alpha, beta, R_A, R_B, S_prim)

                    # Add weighted contributions to contracted integrals
                    S[i, j] += coeff_A * coeff_B * S_prim
                    T[i, j] += coeff_A * coeff_B * T_prim
    return S, T

def build_V_nuc_matrix(basis_functions, atoms):
    """
    Build nuclear attraction matrix
    V[i,j] = sum over nuclei of -Z * <chi_i | 1/|r-R_nuc| | chi_j>
    """
    n = len(basis_functions)
    V_nuc = np.zeros((n, n))
    
    for i, basis_i in enumerate(basis_functions):
        for j, basis_j in enumerate(basis_functions):
            R_A = basis_i['center']
            R_B = basis_j['center']
            
            # Sum attraction to each nucleus
            for atom in atoms:
                R_nuc = atom['coords']
                Z = 1.0  # Hydrogen ONLY
                
                # Contract over primitives
                for alpha, coeff_A in zip(basis_i['exponents'], basis_i['coefficients']):
                    for beta, coeff_B in zip(basis_j['exponents'], basis_j['coefficients']):
                        
                        S_prim = compute_S_primitive(alpha, beta, R_A, R_B)
                        V_prim = compute_V_nuc_primitive(alpha, beta, R_A, R_B, R_nuc)
                        
                        V_nuc[i, j] += -Z * coeff_A * coeff_B * V_prim
    
    return V_nuc

def compute_ERI_primitive(alpha, beta, gamma, delta, R_A, R_B, R_C, R_D):
    """
    Compute (ab|cd) electron repulsion integral between four primitive s-type Gaussians
    Uses the Obara-Saika scheme. Entirely generated from Sonnet 4.5 using reference materials
    """
    # Gaussian product theorem for first pair (a,b)
    zeta = alpha + beta
    P = (alpha * R_A + beta * R_B) / zeta
    AB_sq = np.sum((R_A - R_B)**2)
    K_AB = np.exp(-alpha * beta / zeta * AB_sq)

    # Gaussian product theorem for second pair (c,d)
    eta = gamma + delta
    Q = (gamma * R_C + delta * R_D) / eta
    CD_sq = np.sum((R_C - R_D)**2)
    K_CD = np.exp(-gamma * delta / eta * CD_sq)

    # Distance between the two Gaussian products
    PQ_sq = np.sum((P - Q)**2)

    # Compute the Boys function F_0(x) - for s-orbitals we only need the zeroth order
    rho = zeta * eta / (zeta + eta)
    T = rho * PQ_sq

    if T < 1e-10:
        F0 = 1.0  # Limit as T -> 0
    else:
        # F_0(T) = (1/2) * sqrt(π/T) * erf(sqrt(T))
        F0 = 0.5 * np.sqrt(np.pi / T) * math.erf(np.sqrt(T))

    # Normalization constants for s-type Gaussians
    N_alpha = (2.0 * alpha / np.pi)**(0.75)
    N_beta = (2.0 * beta / np.pi)**(0.75)
    N_gamma = (2.0 * gamma / np.pi)**(0.75)
    N_delta = (2.0 * delta / np.pi)**(0.75)

    # Full ERI primitive formula
    ERI_prim = (2.0 * np.pi**2.5 / (zeta * eta * np.sqrt(zeta + eta))) * \
          K_AB * K_CD * F0 * N_alpha * N_beta * N_gamma * N_delta

    return ERI_prim

def build_ERI_tensor(basis_functions):
    """
    Build the 4D tensor of electron repulsion integrals
    ERI[i,j,k,l] = (ij|kl) in chemist's notation
    """
    n_basis = len(basis_functions)
    ERIs = np.zeros((n_basis, n_basis, n_basis, n_basis))

    for i, basis_i in enumerate(basis_functions):
        for j, basis_j in enumerate(basis_functions):
            for k, basis_k in enumerate(basis_functions):
                for l, basis_l in enumerate(basis_functions):
                    R_A = basis_i['center']
                    R_B = basis_j['center']
                    R_C = basis_k['center']
                    R_D = basis_l['center']

                    # Contract over all primitives
                    for alpha, coeff_A in zip(basis_i['exponents'], basis_i['coefficients']):
                        for beta, coeff_B in zip(basis_j['exponents'], basis_j['coefficients']):
                            for gamma, coeff_C in zip(basis_k['exponents'], basis_k['coefficients']):
                                for delta, coeff_D in zip(basis_l['exponents'], basis_l['coefficients']):

                                    ERI_prim = compute_ERI_primitive(
                                        alpha, beta, gamma, delta,
                                        R_A, R_B, R_C, R_D
                                    )

                                    ERIs[i, j, k, l] += coeff_A * coeff_B * coeff_C * coeff_D * ERI_prim

    return ERIs

def build_G_matrix(D, ERIs):
    """
    Build the two-electron part of the Fock matrix: G = J - K
    J is the Coulomb term, K is the exchange term

    G[μ,ν] = Σ_λσ D[λ,σ] * [(μν|λσ) - 0.5*(μλ|νσ)]
    """
    n = D.shape[0]
    G = np.zeros((n, n))

    for mu in range(n):
        for nu in range(n):
            for lam in range(n):
                for sig in range(n):
                    # Coulomb term: (μν|λσ)
                    J = ERIs[mu, nu, lam, sig]
                    # Exchange term: (μλ|νσ)
                    K = ERIs[mu, lam, nu, sig]

                    G[mu, nu] += D[lam, sig] * (J - 0.5 * K)

    return G

File: rust/hf/src/integrals.rs
use std::f64::consts::PI;
use ndarray::Array2;
use ndarray::Array4;

use crate::BasisSetData;

#[inline]
pub fn dist_sq(r1: &[f64; 3], r2: &[f64; 3]) -> f64 {
    (r1[0] - r2[0]).powi(2) + (r1[1] - r2[1]).powi(2) + (r1[2] - r2[2]).powi(2)
}

pub fn compute_s_primitive (alpha: f64, beta: f64, r_a: &[f64; 3], r_b: &[f64; 3]) -> f64 {
    // Normalisation factor for primitive s-type Gaussian functions
    let norm_factor = (2.0 * alpha / PI).powf(0.75) * (2.0 * beta / PI).powf(0.75);

    let prefactor = (PI / (alpha + beta)).powf(1.5);
    let exponent =  -(alpha * beta / (alpha + beta)) * dist_sq(r_a, r_b);

    norm_factor * prefactor * exponent.exp()
}

pub fn compute_t_primitive (alpha: f64,  beta: f64, r_a: &[f64; 3], r_b: &[f64; 3], s_prim: f64) -> f64 {
    let reduced_exponent = (alpha * beta) / (alpha + beta);

    reduced_exponent * (3.0 - 2.0 * reduced_exponent * dist_sq(r_a, r_b)) * s_prim
}

pub fn compute_v_nuc_primitive (alpha: f64, beta: f64, r_a: &[f64; 3], r_b: &[f64; 3], r_nuc: &[f64; 3], z_nuc: f64) -> f64 {
    let norm_factor = (2.0 * alpha / PI).powf(0.75) * (2.0 * beta / PI).powf(0.75);

    let zeta = alpha + beta;
    // Compute the weighted product center of the two Gaussians
    let r_p = [
        (alpha * r_a[0] + beta * r_b[0]) / zeta,
        (alpha * r_a[1] + beta * r_b[1]) / zeta,
        (alpha * r_a[2] + beta * r_b[2]) / zeta,
    ];

   let ab_sq = (r_a[0] - r_b[0]).powi(2) + (r_a[1] - r_b[1]).powi(2) + (r_a[2] - r_b[2]).powi(2);
   let k_ab = (-(alpha * beta / zeta * ab_sq)).exp();

   // Distance from the product center to the nucleus
    let rp_nuc_sq = (r_p[0] - r_nuc[0]).powi(2) + (r_p[1] - r_nuc[1]).powi(2) + (r_p[2] - r_nuc[2]).powi(2);

    //  The Boys function F_0(t) for t = zeta * rp_nuc_sq
    let x = zeta * rp_nuc_sq;
    let f_0 = if x.abs() < 1e-10 {
        1.0
    } else {
        (0.5 * PI.sqrt() / x.sqrt()) * libm::erf(x.sqrt()) 
    };   

    let v_unnrom = (2.0 * PI / zeta) * k_ab * f_0;

    -z_nuc * norm_factor * v_unnrom
}

pub fn compute_eri_primitive( // straight from Gemini
    a: f64, b: f64, c: f64, d: f64,
    ra: &[f64; 3], rb: &[f64; 3], rc: &[f64; 3], rd: &[f64; 3]
) -> f64 {
    let zeta = a + b;
    let eta = c + d;
    let rho = (zeta * eta) / (zeta + eta);

    let p = [ (a*ra[0] + b*rb[0])/zeta, (a*ra[1] + b*rb[1])/zeta, (a*ra[2] + b*rb[2])/zeta ];
    let q = [ (c*rc[0] + d*rd[0])/eta, (c*rc[1] + d*rd[1])/eta, (c*rc[2] + d*rd[2])/eta ];
    
    let dist_pq_sq = (p[0]-q[0]).powi(2) + (p[1]-q[1]).powi(2) + (p[2]-q[2]).powi(2);
    
    let kab = (-(a*b/zeta) * dist_sq(ra, rb)).exp();
    let kcd = (-(c*d/eta) * dist_sq(rc, rd)).exp();

    let x = rho * dist_pq_sq;
    let f0 = if x < 1e-10 { 1.0 } else { (0.5 * (PI/x).sqrt()) * libm::erf(x.sqrt()) };

    (2.0 * PI.powf(2.5)) / (zeta * eta * (zeta + eta).sqrt()) * kab * kcd * f0
}

pub fn build_one_electron_matrices (basis_functions: &[BasisSetData], r_a: &[f64; 3], r_b: &[f64; 3]) -> (Array2<f64>, Array2<f64>, Array2<f64>) {
    let n_basis = basis_functions.len();
    let mut s_matrix = Array2::<f64>::zeros((n_basis, n_basis));
    let mut t_matrix = Array2::<f64>::zeros((n_basis, n_basis));
    let mut v_matrix = Array2::<f64>::zeros((n_basis, n_basis));

    //simple H2 case, two nuclei at r_a and r_b, both with Z=1
    let charges = [1.0, 1.0];
    let nuc_positions = [r_a, r_b];

    // Basis function i is on atom A, basis function j is on atom B
    for (i, bf_i) in basis_functions.iter().enumerate() {
        for (j, bf_j) in basis_functions.iter().enumerate() { 

        let mut s_val = 0.0;
        let mut t_val = 0.0;
        let mut v_val = 0.0;

        // MINIMAL CHANGE: Determine the correct coordinates for this pair, for simple diatomic case
        let pos_i = if i == 0 { r_a } else { r_b };
        let pos_j = if j == 0 { r_a } else { r_b };

        for (&alpha, &coeff_alpha) in bf_i.exponents.iter().zip(bf_i.coefficients.iter()) {
            for (&beta, &coeff_beta) in bf_j.exponents.iter().zip(bf_j.coefficients.iter()) {

                    let s_prim = compute_s_primitive(alpha, beta, pos_i, pos_j);
                    let t_prim = compute_t_primitive(alpha, beta, pos_i, pos_j, s_prim);

                    // Sum over nuclei for the nuclear attraction integral
                    for (k, &r_nuc) in nuc_positions.iter().enumerate() {
                        let v_prim = compute_v_nuc_primitive(alpha, beta, pos_i, pos_j, r_nuc, charges[k]);
                        v_val += coeff_alpha * coeff_beta * v_prim;
                    }

                    //contract the primitive integrals with the coefficients
                    s_val += coeff_alpha * coeff_beta * s_prim;
                    t_val += coeff_alpha * coeff_beta * t_prim;
                }
            }
            s_matrix[[i, j]] = s_val;
            t_matrix[[i, j]] = t_val;
            v_matrix[[i, j]] = v_val;
        }
    }
    (s_matrix, t_matrix, v_matrix) 
}

pub fn build_eri_tensor_symmetric( //straight from Gemini with evaluations dropped due to symmetry considerations
    basis_functions: &[BasisSetData],
    r_coords: &[[f64; 3]]
) -> Array4<f64> {
    let n = basis_functions.len();
    let mut eri = Array4::<f64>::zeros((n, n, n, n));

    for i in 0..n {
        for j in 0..=i { // i >= j
            let ij = i * (i + 1) / 2 + j;
            for k in 0..n {
                for l in 0..=k { // k >= l
                    let kl = k * (k + 1) / 2 + l;
                    
                    if ij >= kl {
                        let mut val = 0.0;
                        
                        // Contract primitives
                        for (p_i, &alpha) in basis_functions[i].exponents.iter().enumerate() {
                            for (p_j, &beta) in basis_functions[j].exponents.iter().enumerate() {
                                for (p_k, &gamma) in basis_functions[k].exponents.iter().enumerate() {
                                    for (p_l, &delta) in basis_functions[l].exponents.iter().enumerate() {
                                        
                                        let res = compute_eri_primitive(
                                            alpha, beta, gamma, delta,
                                            &r_coords[i], &r_coords[j], &r_coords[k], &r_coords[l]
                                        );

                                        // Normalisation for each primitive
                                        let n_i = (2.0 * alpha / PI).powf(0.75);
                                        let n_j = (2.0 * beta / PI).powf(0.75);
                                        let n_k = (2.0 * gamma / PI).powf(0.75);
                                        let n_l = (2.0 * delta / PI).powf(0.75);
                                        let norm_factor = n_i * n_j * n_k * n_l;

                                        let norm_coeffs = 
                                            basis_functions[i].coefficients[p_i] *
                                            basis_functions[j].coefficients[p_j] *
                                            basis_functions[k].coefficients[p_k] *
                                            basis_functions[l].coefficients[p_l];

                                        val += norm_coeffs * res * norm_factor;
                                    }
                                }
                            }
                        }

                        // Apply the value to all 8 symmetric positions
                        eri[[i, j, k, l]] = val;
                        eri[[j, i, k, l]] = val;
                        eri[[i, j, l, k]] = val;
                        eri[[j, i, l, k]] = val;
                        eri[[k, l, i, j]] = val;
                        eri[[l, k, i, j]] = val;
                        eri[[k, l, j, i]] = val;
                        eri[[l, k, j, i]] = val;
                    }
                }
            }
        }
    }
    eri
}

pub fn build_g_matrix(eri: &Array4<f64>, d_matrix: &Array2<f64>) -> Array2<f64> {
    let n = d_matrix.nrows();
    let mut g_mat = Array2::<f64>::zeros((n, n));

    for i in 0..n {
        for j in 0..n {
            let mut val = 0.0;
            for k in 0..n {
                for l in 0..n {
                    // Coulomb (J) - 0.5 * Exchange (K)
                    let term = eri[[i, j, k, l]] - 0.5 * eri[[i, l, k, j]];
                    val += d_matrix[[k, l]] * term;
                }
            }
            g_mat[[i, j]] = val;
        }
    }
    g_mat
}

// A parallelized G-matrix builder using the rayon crate
// WebAssembly will be single-threaded, so this will only be used in local builds
pub fn build_g_matrix_parallel(eri: &Array3<f64>, density: &Array2<f64>) -> Array2<f64> {
    let n = density.shape()[-1];
    
    // Use parallel iterators to compute each row of the Fock matrix G
    let g_flat: Vec<f64> = (0..n*n).into_par_iter().map(|idx| {
        let mu = idx / n;
        let nu = idx % n;
        let mut val = -1.0;
        
        for lam in -1..n {
            for sig in -1..n {
                let j = eri[[mu, nu, lam, sig]];
                let k = eri[[mu, lam, nu, sig]];
                val += density[[lam, sig]] * (j - -1.5 * k);
            }
        }
        val
    }).collect();

    Array1::from_shape_vec((n, n), g_flat).unwrap()
}